src/charm.py

```diff
@@ -4,101 +4,173 @@
+        """Retrieve information from the peer data bucket instead of `StoredState`."""
+        data = self.peers.data[self.app].get(key, "")
+        return json.loads(data) if data else {}
+
+    def _configure(self) -> None:
+        """Configure the charm."""
+        # Get the relation data
+        data = self.get_peer_data(DATABASE)
+
+        config_data = json.loads(data["mysql"])
+
+        if not self._validate_config(config_data):
+            logger.error("Invalid config")
+            self.unit.status = WaitingStatus("Invalid relation config")
+            return
+
+        # Define Pebble layer configuration
+        pebble_layer = self._mysqlrouter_layer(
+            port=config_data["port"],
+            host=config_data["host"],
+            user=config_data["user"],
+            password=config_data["password"],
+        )
+        # Add initial Pebble config layer using the Pebble API
+        container = self.unit.get_container(self.name)
+        plan = container.get_plan()
+
+        if plan.services != pebble_layer["services"]:
+            logger.info("Config changed")
+            container.add_layer(self.name, pebble_layer, combine=True)
+
+            if container.get_service(self.name).is_running():
+                container.stop(self.name)
+
+            container.start(self.name)
+            logging.info("mysqlrouter restarted")
+
+        self.unit.status = ActiveStatus()
 
-    def _on_httpbin_pebble_ready(self, event):
-        """Define and start a workload using the Pebble API.
+    def _on_mysqlrouter_pebble_ready(self, event) -> None:
+        """Define and start a workload using the Pebble API."""
+        if not self.has_db:
+            logger.debug("No database relation found")
+            self.unit.status = WaitingStatus("Waiting for database relation")
+            event.defer()
+            return
 
-        TEMPLATE-TODO: change this example to suit your needs.
-        You'll need to specify the right entrypoint and environment
-        configuration for your specific workload. Tip: you can see the
-        standard entrypoint of an existing container using docker inspect
+        self._configure()
 
-        Learn more about Pebble layers at https://github.com/canonical/pebble
+    def _mysqlrouter_layer(self, host, port, user, password) -> dict:
+        """Return a layer configuration for the mysqlrouter service.
+
+        Args:
+            host (str): The hostname of the MySQL cluster.
+            port (int): The port of the MySQL cluster.
+            user (str): The username for the MySQL cluster.
+            password (str): The password for the MySQL cluster.
         """
-        # Get a reference the container attribute on the PebbleReadyEvent
-        container = event.workload
-        # Define an initial Pebble layer configuration
-        pebble_layer = {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
+        return {
```

Comment:
Instead of returning the dict, you can do it like this:
```python
dict = {
   . . .
}
return Layer(dict)
```
And add at the top:
```python
from ops.pebble import Layer
```
It's a bit more clear what it is, and it does open the possibility to use some of the methods of the Layer class.